OSGi java.lang.ClassNotFoundException: com.hazelcast.hibernate.instance.DefaultHazelcastInstanceFactory

Hi,
I am trying to use hazelcast-hibernate5-1.2.2 in an OSGi environment but I get the following exception on startup:

> Caused by: org.apache.felix.log.LogException: org.hibernate.cache.CacheException: Failed to set up hazelcast instance factory
> 	at com.hazelcast.hibernate.AbstractHazelcastCacheRegionFactory.start(AbstractHazelcastCacheRegionFactory.java:91)
> 	at org.hibernate.internal.CacheImpl.<init>(CacheImpl.java:49)
> 	at org.hibernate.engine.spi.CacheInitiator.initiateService(CacheInitiator.java:28)
> 	at org.hibernate.engine.spi.CacheInitiator.initiateService(CacheInitiator.java:20)
> 	at org.hibernate.service.internal.SessionFactoryServiceRegistryImpl.initiateService(SessionFactoryServiceRegistryImpl.java:46)
> 	at org.hibernate.service.internal.AbstractServiceRegistryImpl.createService(AbstractServiceRegistryImpl.java:234)
> 	... 16 more
> Caused by: java.lang.ClassNotFoundException: com.hazelcast.hibernate.instance.DefaultHazelcastInstanceFactory
> 	at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1284)
> 	at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1118)
> 	at java.lang.Class.forName0(Native Method)
> 	at java.lang.Class.forName(Class.java:348)
> 	at com.hazelcast.hibernate.AbstractHazelcastCacheRegionFactory.start(AbstractHazelcastCacheRegionFactory.java:88)

The problem is the code in method `AbstractHazelcastCacheRegionFactory.start(...)` which uses `Thread.currentThread().getContextClassLoader()` to try to load the class `DefaultHazelcastInstanceFactory` using `Class.forName`. 

This is unnecessary as both classes are in the same bundle so just calling `new DefaultHazelcastInstanceFactory()` would be enough in the case that no property `hibernate.cache.hazelcast.factory` is defined. 

Using the Thread's ContextClassLoader in an OSGi environment is generally not a good idea, it [seems](http://blog.osgi.org/2011/05/what-you-should-know-about-class.html).

When hibernate tries to read data from the cache I get this exception. I have to analyze this further.

```
2018-06-26 15:56:09,474 ERROR [hazelcast.event-3] .impl.eventservice.impl.EventServiceImpl - [127.0.0.1]:5701 [3.9] hz.test-hazelcast.event-3 caught an exception while processing task:com.hazelcast.spi.impl.eventservice.impl.LocalEventDispatcher@900b7b7
com.hazelcast.nio.serialization.HazelcastSerializationException: java.lang.ClassNotFoundException: org.hibernate.engine.jdbc.Size
	at com.hazelcast.internal.serialization.impl.JavaDefaultSerializers$JavaSerializer.read(JavaDefaultSerializers.java:224)
	at com.hazelcast.internal.serialization.impl.StreamSerializerAdapter.read(StreamSerializerAdapter.java:48)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.readObject(AbstractSerializationService.java:267)
	at com.hazelcast.internal.serialization.impl.ByteArrayObjectDataInput.readObject(ByteArrayObjectDataInput.java:570)
	at com.hazelcast.hibernate.region.CacheKeyImpl.readData(CacheKeyImpl.java:89)
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.readInternal(DataSerializableSerializer.java:158)
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.read(DataSerializableSerializer.java:105)
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.read(DataSerializableSerializer.java:50)
	at com.hazelcast.internal.serialization.impl.StreamSerializerAdapter.read(StreamSerializerAdapter.java:48)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.readObject(AbstractSerializationService.java:267)
	at com.hazelcast.internal.serialization.impl.ByteArrayObjectDataInput.readObject(ByteArrayObjectDataInput.java:570)
	at com.hazelcast.hibernate.local.Invalidation.readData(Invalidation.java:59)
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.readInternal(DataSerializableSerializer.java:158)
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.read(DataSerializableSerializer.java:105)
	at com.hazelcast.internal.serialization.impl.DataSerializableSerializer.read(DataSerializableSerializer.java:50)
	at com.hazelcast.internal.serialization.impl.StreamSerializerAdapter.read(StreamSerializerAdapter.java:48)
	at com.hazelcast.internal.serialization.impl.AbstractSerializationService.toObject(AbstractSerializationService.java:185)
	at com.hazelcast.topic.impl.DataAwareMessage.getMessageObject(DataAwareMessage.java:44)
	at com.hazelcast.hibernate.local.LocalRegionCache$1.onMessage(LocalRegionCache.java:208)
	at com.hazelcast.topic.impl.TopicService.dispatchEvent(TopicService.java:136)
	at com.hazelcast.spi.impl.eventservice.impl.LocalEventDispatcher.run(LocalEventDispatcher.java:64)
	at com.hazelcast.util.executor.StripedExecutor$Worker.process(StripedExecutor.java:225)
	at com.hazelcast.util.executor.StripedExecutor$Worker.run(StripedExecutor.java:208)
Caused by: java.lang.ClassNotFoundException: org.hibernate.engine.jdbc.Size
	at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1284)
	at org.apache.catalina.loader.WebappClassLoaderBase.loadClass(WebappClassLoaderBase.java:1118)
	at com.hazelcast.nio.ClassLoaderUtil.tryLoadClass(ClassLoaderUtil.java:173)
	at com.hazelcast.nio.ClassLoaderUtil.loadClass(ClassLoaderUtil.java:147)
	at com.hazelcast.nio.IOUtil$ClassLoaderAwareObjectInputStream.resolveClass(IOUtil.java:552)
	at java.io.ObjectInputStream.readNonProxyDesc(ObjectInputStream.java:1859)
```

--- a/hazelcast-hibernate5/src/main/java/com/hazelcast/hibernate/AbstractHazelcastCacheRegionFactory.java
+++ b/hazelcast-hibernate5/src/main/java/com/hazelcast/hibernate/AbstractHazelcastCacheRegionFactory.java
@@ -32,6 +32,7 @@ import org.hibernate.cache.spi.access.AccessType;
 import java.util.Properties;
 
 import static com.hazelcast.hibernate.CacheEnvironment.getCacheCleanup;
+import static java.lang.Class.forName;
 
 /**
  * Abstract superclass of Hazelcast based {@link RegionFactory} implementations
@@ -81,20 +82,26 @@ public abstract class AbstractHazelcastCacheRegionFactory implements RegionFacto
     public void start(final SessionFactoryOptions options, final Properties properties) throws CacheException {
         log.info("Starting up " + getClass().getSimpleName());
         if (instance == null || !instance.getLifecycleService().isRunning()) {
-            String defaultFactory = DefaultHazelcastInstanceFactory.class.getName();
-            String factoryName = properties.getProperty(CacheEnvironment.HAZELCAST_FACTORY, defaultFactory);
-            ClassLoader cl = Thread.currentThread().getContextClassLoader();
-            try {
-                Class<IHazelcastInstanceFactory> factory =
-                        (Class<IHazelcastInstanceFactory>) Class.forName(factoryName, true, cl);
-                instanceLoader = factory.newInstance().createInstanceLoader(properties);
-            } catch (ClassNotFoundException | IllegalAccessException | InstantiationException e) {
-                throw new CacheException("Failed to set up hazelcast instance factory", e);
-            }
+            instanceLoader = resolveInstanceLoader(properties);
             instance = instanceLoader.loadInstance();
         }
 
         cleanupService = new CleanupService(instance.getName(), getCacheCleanup(properties));
+    }
+
+    private IHazelcastInstanceLoader resolveInstanceLoader(Properties properties) {
+        String factoryName = properties.getProperty(CacheEnvironment.HAZELCAST_FACTORY);
+        if (factoryName != null) {
+            ClassLoader cl = Thread.currentThread().getContextClassLoader();
+            try {
+                Class<IHazelcastInstanceFactory> factory = (Class<IHazelcastInstanceFactory>) forName(factoryName, true, cl);
+                return factory.newInstance().createInstanceLoader(properties);
+            } catch (ClassNotFoundException | IllegalAccessException | InstantiationException e) {
+                throw new CacheException("Failed to set up hazelcast instance factory", e);
+            }
+        } else {
+            return new DefaultHazelcastInstanceFactory().createInstanceLoader(properties);
+        }
     }
 
     @Override

--- a/hazelcast-hibernate52/src/main/java/com/hazelcast/hibernate/AbstractHazelcastCacheRegionFactory.java
+++ b/hazelcast-hibernate52/src/main/java/com/hazelcast/hibernate/AbstractHazelcastCacheRegionFactory.java
@@ -32,6 +32,7 @@ import org.hibernate.cache.spi.access.AccessType;
 import java.util.Properties;
 
 import static com.hazelcast.hibernate.CacheEnvironment.getCacheCleanup;
+import static java.lang.Class.forName;
 
 /**
  * Abstract superclass of Hazelcast based {@link RegionFactory} implementations
@@ -81,20 +82,26 @@ public abstract class AbstractHazelcastCacheRegionFactory implements RegionFacto
     public void start(final SessionFactoryOptions options, final Properties properties) throws CacheException {
         log.info("Starting up " + getClass().getSimpleName());
         if (instance == null || !instance.getLifecycleService().isRunning()) {
-            String defaultFactory = DefaultHazelcastInstanceFactory.class.getName();
-            String factoryName = properties.getProperty(CacheEnvironment.HAZELCAST_FACTORY, defaultFactory);
-            ClassLoader cl = Thread.currentThread().getContextClassLoader();
-            try {
-                Class<IHazelcastInstanceFactory> factory =
-                        (Class<IHazelcastInstanceFactory>) Class.forName(factoryName, true, cl);
-                instanceLoader = factory.newInstance().createInstanceLoader(properties);
-            } catch (ClassNotFoundException | IllegalAccessException | InstantiationException e) {
-                throw new CacheException("Failed to set up hazelcast instance factory", e);
-            }
+            instanceLoader = resolveInstanceLoader(properties);
             instance = instanceLoader.loadInstance();
         }
 
         cleanupService = new CleanupService(instance.getName(), getCacheCleanup(properties));
+    }
+
+    private IHazelcastInstanceLoader resolveInstanceLoader(Properties properties) {
+        String factoryName = properties.getProperty(CacheEnvironment.HAZELCAST_FACTORY);
+        if (factoryName != null) {
+            ClassLoader cl = Thread.currentThread().getContextClassLoader();
+            try {
+                Class<IHazelcastInstanceFactory> factory = (Class<IHazelcastInstanceFactory>) forName(factoryName, true, cl);
+                return factory.newInstance().createInstanceLoader(properties);
+            } catch (ClassNotFoundException | IllegalAccessException | InstantiationException e) {
+                throw new CacheException("Failed to set up hazelcast instance factory", e);
+            }
+        } else {
+            return new DefaultHazelcastInstanceFactory().createInstanceLoader(properties);
+        }
     }
 
     @Override

--- a/hazelcast-hibernate53/src/main/java/com/hazelcast/hibernate/AbstractHazelcastCacheRegionFactory.java
+++ b/hazelcast-hibernate53/src/main/java/com/hazelcast/hibernate/AbstractHazelcastCacheRegionFactory.java
@@ -38,8 +38,8 @@ import org.hibernate.engine.spi.SessionFactoryImplementor;
 import java.util.Map;
 import java.util.Properties;
 
-import static com.hazelcast.hibernate.CacheEnvironment.HAZELCAST_FACTORY;
 import static com.hazelcast.hibernate.CacheEnvironment.getCacheCleanup;
+import static java.lang.Class.forName;
 
 /**
  * Simple RegionFactory implementation to return Hazelcast based local Region implementations
@@ -133,30 +133,30 @@ public abstract class AbstractHazelcastCacheRegionFactory extends RegionFactoryT
         return super.isStarted() && instance.getLifecycleService().isRunning();
     }
 
-    @SuppressWarnings("unchecked")
     @Override
     protected void prepareForUse(final SessionFactoryOptions settings, final Map configValues) {
         log.info("Starting up " + getClass().getSimpleName());
         if (instance == null || !instance.getLifecycleService().isRunning()) {
-            final String defaultFactory = DefaultHazelcastInstanceFactory.class.getName();
-
-            String factoryName = (String) configValues.get(HAZELCAST_FACTORY);
-            if (factoryName == null) {
-                factoryName = defaultFactory;
-            }
-
-            final ClassLoader cl = Thread.currentThread().getContextClassLoader();
-            try {
-                final Class<IHazelcastInstanceFactory> factory =
-                  (Class<IHazelcastInstanceFactory>) Class.forName(factoryName, true, cl);
-                instanceLoader = factory.newInstance().createInstanceLoader(toProperties(configValues));
-            } catch (ClassNotFoundException | InstantiationException | IllegalAccessException e) {
-                throw new CacheException("Failed to set up hazelcast instance factory", e);
-            }
+            instanceLoader = resolveInstanceLoader(toProperties(configValues));
             instance = instanceLoader.loadInstance();
         }
 
         cleanupService = new CleanupService(instance.getName(), getCacheCleanup(toProperties(configValues)));
+    }
+
+    private IHazelcastInstanceLoader resolveInstanceLoader(Properties properties) {
+        String factoryName = properties.getProperty(CacheEnvironment.HAZELCAST_FACTORY);
+        if (factoryName != null) {
+            ClassLoader cl = Thread.currentThread().getContextClassLoader();
+            try {
+                Class<IHazelcastInstanceFactory> factory = (Class<IHazelcastInstanceFactory>) forName(factoryName, true, cl);
+                return factory.newInstance().createInstanceLoader(properties);
+            } catch (ClassNotFoundException | IllegalAccessException | InstantiationException e) {
+                throw new CacheException("Failed to set up hazelcast instance factory", e);
+            }
+        } else {
+            return new DefaultHazelcastInstanceFactory().createInstanceLoader(properties);
+        }
     }
 
     @SuppressWarnings("Duplicates")

